### ISSUE TITLE
OGR Errors with CONNECTIONTYPE OGR and WFS 2.0

### ISSUE THREAD

**Issue by geographika** [MapServer/MapServer#6155]
Requesting GetFeature on a Shapefile with an OGR CONNECTIONTYPE returns the following error - but only when requesting through a web server, and using WFS 2.0.0. Using WFS 1.1.0 works fine:

```
msWFSGetFeature(): WFS server error. ms_error-&gt;code not found
msOGRFileNextShape(): OGR error. OGR GetNextFeature() error'd. Check logs
```

Thrown at https://github.com/MapServer/MapServer/blob/d91fec44118f7c421355e8632d20bd96c7152887/mapogr.cpp#L2796

Debugging notes and details:

- using a master build of MapServer 7.7-dev and GDAL 3.1.0.0 (GISInternals SDK)
- running on the command line is fine and returns GML without errors - sample command: `mapserv -nh "QUERY_STRING=map=example2.map&SERVICE=WFS&VERSION=2.0.0&REQUEST=GetFeature&typeName=Cemeteries"`
- running on a web server (IIS) causes the errors. Sample request: http://localhost/mapserver2/?map=D:/Github/mapview-demo/example2.map&service=WFS&version=2.0.0&request=GetFeature&typeName=cemeteries
- if the number of features to return is limited to the number of features in the shapefile then valid results are returned. E.g. if there are 100 features in the Shapefile:

  - the following will work: http://localhost/mapserver2/?map=D:/Github/mapview-demo/example2.map&service=WFS&version=2.0.0&request=GetFeature&typeName=cemeteries&startIndex=0&count=100
  - the following will trigger the error: http://localhost/mapserver2/?map=D:/Github/mapview-demo/example2.map&service=WFS&version=2.0.0&request=GetFeature&typeName=cemeteries&startIndex=0&count=101

- same error on different Shapefiles. All features in the test shapefiles appear valid
- OUTPUTFORMAT makes no difference

Minimal Mapfile below:

```
MAP
    NAME "MapviewDemo"
    EXTENT -1981303.751 5648520.031 522771.571 8580038.317
    CONFIG "PROJ_LIB" "C:\MapServer\bin\proj\SHARE"
    SIZE 400 300
    WEB
      METADATA
        "ows_enable_request" "*"
        "wfs_srs" "EPSG:3857 EPSG:2157 EPSG:29902 EPSG:4326 EPSG:3978"
        "wfs_onlineresource" "http://localhost/mapserver2/?map=/MapServer/apps/mapview-demo/example.map&"
      END
    END

    PROJECTION
      "init=epsg:3857"
    END

    LAYER
      NAME "cemeteries"

      CONNECTIONTYPE OGR
      CONNECTION 'data/cemeteries.shp'

      PROJECTION
        "init=epsg:3857"
      END
      TYPE POLYGON
      METADATA
          "gml_include_items" "all"
          "gml_types" "auto"
          "gml_featureid" "osm_id"
      END          
      CLASS
        NAME 'Cemeteries'
        STYLE
          COLOR 244 123 233
        END
      END
    END

END
```

I posted this error before at http://osgeo-org.1560.x6.nabble.com/Filter-could-not-be-translated-to-OGR-filter-td5432557.html but believed it was related to filters. It does not appear this is the case. 

I think it must be related to feature counts and requesting feature indexes higher than the count. Alternatively maybe related to threads/FastCGI and missing a `CPLErrorReset()`.

**Comment by geographika** [comment]
Not sure how I missed it, but in the logs was:

```
[Sat Oct  3 01:31:22 2020].404000 CGI Request 1 on process 503108
[Sat Oct  3 01:31:24 2020].125000 msOGRFileNextShape(): 'PROJ: proj_identify: Cannot find proj.db'
[Sat Oct  3 01:31:24 2020].125000 msOGRFileNextShape(): OGR error. OGR GetNextFeature() error'd. Check logs.
```

The proj.db was in C:\MapServer\bin\proj\SHARE = changing to C:\MapServer\bin\proj6\share any copying in the epsg file avoided this error and allowed the request to finish. 
I'm not sure how the number of records returned could affect this, or why WFS 2.0 works differently. 

There still seem to be some issues with filters however.

**Comment by geographika** [comment]
After further debugging, the following line is where the error is set:

https://github.com/MapServer/MapServer/blob/d91fec44118f7c421355e8632d20bd96c7152887/mapogr.cpp#L2793

The error is set in PROJ (6.1):

`PROJ: proj_identify: Cannot find proj.db`

The code runs fine but then after all features have been looped through there is a check for any errors and `MS_FAILURE` is returned:

https://github.com/MapServer/MapServer/blob/d91fec44118f7c421355e8632d20bd96c7152887/mapogr.cpp#L2795

I'm not sure what, if anything, should be done. If the code runs fine even when proj.db is not found then it doesn't seem like it should fail on the request.

**Comment by rouault** [comment]
> any copying in the epsg file

you shouldn't do that, unless you've put custom entries in it. If the epsg file is place in \proj\share, this will prevent proj.db to be used when resolving EPSG code, and will fallback to the epsg file instead.

**Comment by geographika** [comment]
There seems to be a clash between the proj.db and epsg file somewhere. Setting the following at the top of the Mapfile:

`CONFIG "PROJ_LIB" "C:/MapServer/bin/proj/SHARE"`

Causes the errors in this issue with not returning GetFeature queries:

`msOGRFileNextShape(): PROJ: proj_identify: Cannot find proj.db`

Setting to:

`CONFIG "PROJ_LIB" "C:/MapServer/bin/proj6/SHARE"`

Leads to:

`msProcessProjection(): Projection library error. proj error &quot;No such file or directory&quot; for &quot;init=epsg:3857&quot;`

Copying proj.db into `C:/MapServer/bin/proj/SHARE` or the epsg file into `C:/MapServer/bin/proj6/SHARE` (advised against above) gets the GetFeature to run correctly from the command line but not the web server.

**Comment by geographika** [comment]
Simply setting the following to `MS_DONE` gets rid of all issues with GetFeatures/filters etc. and returns valid results (in my case - if there were other errors then this would be a problem):

https://github.com/MapServer/MapServer/blob/d91fec44118f7c421355e8632d20bd96c7152887/mapogr.cpp#L2801

A correct resolution would be to ignore the `PROJ: proj_identify: Cannot find proj.db` errors if they are not critical, or fix the error by being able to point the PROJ_LIB to both proj.db and the epsg file (which I've not been able to figure out yet).

**Comment by rouault** [comment]
I believe the issue is that CONFIG PROJ_LIB only affects MapServer configuration of PROJ, and not GDAL one. Try to define PROJ_LIB as an environment variable instead, at least to check my hypothesis.

**Comment by rouault** [comment]
or alternatively, with CONFIG PROJ_LIB, could you try the following patch ?
```
diff --git a/mapproject.c b/mapproject.c
index 5659c59..e432ae1 100644
--- a/mapproject.c
+++ b/mapproject.c
@@ -36,6 +36,8 @@
 #include <sys/stat.h>
 #include "mapaxisorder.h"
 
+#include "ogr_srs_api.h"
+
 static char *ms_proj_lib = NULL;
 #if PROJ_VERSION_MAJOR >= 6
 static unsigned ms_proj_lib_change_counter = 0;
@@ -2399,6 +2401,14 @@ void msSetPROJ_LIB( const char *proj_lib, const char *pszRelToPath )
 #endif
   msReleaseLock( TLOCK_PROJ );
 
+#if GDAL_VERSION_MAJOR >= 3
+  if( ms_proj_lib != NULL )
+  {
+    const char* const apszPaths[] = { ms_proj_lib, NULL };
+    OSRSetPROJSearchPaths(apszPaths);
+  }
+#endif
+
   if ( extended_path )
     msFree( extended_path );
 }
```

**Comment by geographika** [comment]
@rouault - I wasn't aware that CONFIG PROJ_LIB wouldn't set the GDAL PROJ, but I think that makes sense as there were errors in the log trying to find files in the `proj6/SHARE` and `proj/SHARE folders` when I set PROJ_LIB as a FastCGI environment variable. I can't set a system env on Windows as these are ignored when running under FastCGI (see [old mailing list post](http://osgeo-org.1560.x6.nabble.com/MapServer-Environment-Variables-and-IIS-td5186363.html)). I couldn't resolve the issue after trying many combinations of these settings. 

The patch above however did fix everything - thank you! Is this copying the MapServer proj_lib path to GDAL/OGR?
I've added as a pull request if you think this should be merged.

**Comment by rouault** [comment]
> Is this copying the MapServer proj_lib path to GDAL/OGR?

yes, this is propagating CONFIG PROJ_LIB to GDAL, so that it uses it when configuring its own PROJ contexts.

> I've added as a pull request if you think this should be merged.

Yes, definitely a fix to merge

**Comment by geographika** [comment]
Fixed with #6156 - thanks @rouault
